Update DuckDB Package to fix Version Determination Bug and Add Variants

teaguesterling opened the pull request:
This revises the DuckDB package in the builtin repo in the following ways:
- Changes the build system from `CMakePackage` to `Makefile`. While DuckDB uses CMake, the build process is documented to begin with make. This also allows `edit()` and `patch()` to be called as needed (unless I misunderstood something).
- Patches the `CMakeLists.txt` file to hard-code the version number from the Spack version instead of trying to compute it from the git tag. Since we're pulling tarballs, there is no git commit information and it defaults to an incorrect number. This, in turn, breaks the usage of external extensions (e.g., `postgres_scanner`) which are tied to a minor release number.
- Adds variants to build the extensions included in the repo (and some semi-opinionated defaults). The only two not built by default are the JDBC and ODBC connectors as I have not worked out the dependencies at the time of writing this.
- Adds two more up-to-date versions of the package and updates the default to the latest release, which is `v0.8.1`.
- Adds Python >=3.10 as a dependency, as is stated in the DuckDB repository.
- Adds a `url_for_version` function for future releases.

teaguesterling committed:
Expanding the duckdb package to fix the version number (required for external extensions to work) being pulled from git and have variants for the built-in extensions at build-time. This also changes the build system from CMakePackage to Makefile package. While DuckDB uses CMake, the build process originates with a Makefile. This has the benefit of letting us use edit/patch as well.

teaguesterling committed:
Conforming to formatting and style guide. Also removing unneeded messages

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
Adding myself as a maintainer of this package

teaguesterling committed:
Merge branch 'packages/duckdb-with-version' of github.com:teaguesterling/spack into packages/duckdb-with-version

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
Update package.py

Adding descriptions to variants as noted by linter in pull request

teaguesterling committed:
Update package.py

Shortening line lengths

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
Merge branch 'spack:develop' into packages/duckdb-with-version

teaguesterling committed:
Merge pull request #1 from spack/develop

Merge

teaguesterling committed:
Adding new versions

teaguesterling committed:
Merge changes to duckdb license

teaguesterling committed:
Adding new versions to duckdb package

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling commented:
@glentner I saw your pull request of the 0.9.2 version and would appreciate your thoughts on this more substantial set of changes for DuckDB. I've had some trouble getting any attention (or rejection) for this PR.

glentner commented:
I also noticed the version number bug but hadn't had any time to dig into Spack to figure out why it was failing. 

This all sounds great and sensible; I'll take a look this week and get back with feedback. 

I want to get the 0.10.0 release in, but think we should get this squared away first. 

Thanks.

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,65 @@ class Duckdb(CMakePackage):
+    variant("json", default=True, description="Include JSON support in build")
+    variant("httpfs", default=True, description="Include HTTPFS (& S3) support in build")
+    variant("excel", default=True, description="Include Excel formatting extension in build")
+    variant("openssl", default=True, description="Compile with bundled OpenSSl library")
```
I think this is no longer there in 0.10.0.

teaguesterling replied:
Updated this in latest commit to be default=False and added explicit dependency on openssl.

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,65 @@ class Duckdb(CMakePackage):
+    variant("inet", default=True, description="Include INET (ip address) support in build")
+    variant("fts", default=True, description="Include FTS (full text search) support in build")
+    variant("parquet", default=True, description="Include parquent support in build")
+
```
Should we add support for python here?

teaguesterling replied:
Added it bu there's more work to do in order to get the various client APIs to build as expected.

teaguesterling replied:
Will mark this resolved for now but further work is needed.

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,65 @@ class Duckdb(CMakePackage):
+            if "+" + flag in spec:
+                env[make_flag] = "1"
+            elif "~" + flag in spec:
+                env[make_flag] = "0"
```
The duckdb docs recommend building with ninja.
Not sure if it makes much of a difference for spack, but it might work if we add a dependency for ninja and then add the "GEN=ninja" environment variable here.

teaguesterling replied:
Works with or without Ninja. Added a (default=True) variant to use ninja as proposed here and set up the necessary dependencies.

EbiArnie reviewed (changes requested):
I have some suggestions, please see inline comments.

teaguesterling committed:
- Reorganized and cleaned up variants
 - Updated the patch to work for 0.10.0+
 - Removed (or made non-default) some unecessary variants
 - Added ninja as the default generator
 - Set up some shared library dependencies;

teaguesterling commented:
The prior filter_file pattern for injecting version into CMakeLists.txt no longer works in 0.10.0 as there is another DUCKDB_NORMALIZED_VERSION that also needs to get set.

teaguesterling committed:
Fixing style and line length issues

teaguesterling committed:
Fixing style issues again

teaguesterling committed:
line extension style was off

teaguesterling committed:
line length was off

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
It seems we need to remove the '-devX' from the version to get extensions to work

teaguesterling committed:
Fix broken patch version replacement

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

teaguesterling committed:
Fix quote style

teaguesterling committed:
line lenght

teaguesterling commented:
Last update here. Extension loading works as expected now after removing "-dev0" from the version string and setting both `DUCKDB_VERSION` and `DUCKDB_NORMALIZED_VERSION`. Tested by installing & loading the postgres_scanner extension into a fresh install:
```
=> duckdb: Successfully installed duckdb-0.10.0-almdysxm2mtsv4wgpm424kevcri2lbkw
  Stage: 1.23s.  Edit: 0.00s.  Build: 9m 8.79s.  Install: 0.03s.  Post-install: 0.16s.  Total: 9m 10.76s
[+] /home/teague/.spack-inst/opt/spack/linux-linuxmint21-ivybridge/gcc-11.3.0/duckdb-0.10.0-almdysxm2mtsv4wgpm424kevcri2lbkw
==> Updating view at /home/teague/.spack-inst/var/spack/environments/duckdb-test/.spack-env/view
~t$ duckdb
v0.10.0 
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D install postgres_scanner;
D load postgres_scanner;
D 
```

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+    variant("cli", default=True, description="Compile with command line client")
+    variant("icu", default=False, description="Compile with bundled ICU library")
+    variant("ninjabuild", default=True, description="Use GEN=ninja to build")
+    variant("openssl", default=False, description="Compile with bundled OpenSSL library")
```
This is no longer a variant in 0.10.0.
This should (probably) become 'variant("openssl", default=False, description="...", when="@:0.9.2")'
, more or less. I did not check the version here.

teaguesterling replied:
Added this into the PR

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+        # We use the Spack version to inject it in the right place during the build
+
+        v = self.spec.version
+        if v < Version("0.10.0"):
```
It is more common to use a "with when():" block here. But since you need the version anyway, I think this is fine.

teaguesterling replied:
Used the suggested change to use satisfies

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+    version("0.9.1", sha256="37a43188d9354ce3ca101b2b118d867f5f76d04c3b83c09d86fd7508351a631b")
+    version("0.9.0", sha256="3dbf3326a831bf0797591572440e81a3d6d668f8e33a25ce04efae19afc3a23d")
     version("0.8.1", sha256="a0674f7e320dc7ebcf51990d7fc1c0e7f7b2c335c08f5953702b5285e6c30694")
+    version("0.8.0", sha256="df3b8e0b72bce38914f0fb1cd02235d8b616df9209beb14beb06bfbcaaf2e97f")
```
Maybe older versions should be marked as deprecated:
version("...", sha256="...", deprecated=True)

teaguesterling replied:
Deprecated all but 0.10.0

EbiArnie reviewed (approved):
This works for me. Built the package and tested with the mysql_scanner extension.
I did not test jdbc, odbc or python. As you said, python needs more work.
I would be happy for this to be merged as-is, it provides the new version and working custom extensions.
I have some nitpicks, they can be addressed later - if at all.

EbiArnie commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+    version("0.8.0", sha256="df3b8e0b72bce38914f0fb1cd02235d8b616df9209beb14beb06bfbcaaf2e97f")
     version("0.7.1", sha256="67f840f861e5ffbe137d65a8543642d016f900b89dd035492d562ad11acf0e1e")
+
+    depends_on("python@3:")
```
The python extension for duckdb needs version 3.7 minimum. I did not find a minimum version for duckdb without the extension, but I would be surprised if all of the included scripts would run with Python 3.0. I think this should be higher.
Since Spack needs minimum 3.6, it is probably safe to just bump this up.

teaguesterling replied:
Agreed. Probably safest to aim for 3.7 without any further guidance from duckdb.

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
 
     license("MIT")
     maintainers("glentner")
+    maintainers("teaguesterling")
```
```suggestion
    maintainers("glentner", "teaguesterling")
```

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+    variant("autocomplete", default=True, description="Include autocomplete for CLI in build")
+    variant("cli", default=True, description="Compile with command line client")
+    variant("icu", default=False, description="Compile with bundled ICU library")
+    variant("ninjabuild", default=True, description="Use GEN=ninja to build")
```
I'm a bit puzzled by the way this is packaged. We use `gmake` to drive `cmake` to generate `ninja` builds :thinking: I get that this is how upstream advise to build from sources... but is there a real drawback in using CMake directly?[^1]

Note that `patch` can be called independently from the base class being selected.

[^1]: Somehow I'm not persuaded by `It is not advised to directly call CMake, as the Makefile sets certain variables that are crucial to properly building the package.` Wondering if you have a bit more info on that.

teaguesterling replied:
It seems that there is some logic in the Makefile that handles multiple cmake builds for supporting extensions, but the default behavior doesn't trigger those additional targets. That said, the Makefile appears to be fairly actively developed. I initially worried that it would be difficult to keep the cmake variable setting in sync, especially around extension handling.

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+    variant("odbc", default=False, description="Build with ODBC driver (may not work)")
+    variant("python", default=False, description="Build with Python driver (may not work)")
+
+    def edit(self, spec, prefix):
+        if "+ninjabuild" in spec:
+            env["GEN"] = "ninja"
```
Can we use `setup_build_environment` to set environment variables?

teaguesterling replied:
Added this change in.

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+        # We use the Spack version to inject it in the right place during the build
+
+        v = self.spec.version
+        if v < Version("0.10.0"):
```
The:
```python
with when(...):
    ...
```
block works only with directives. Within a method like this it's more idiomatic to:
```suggestion
        if not self.spec.satisfies("@0.10.0:"):
```

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
+                "CMakeLists.txt",
+            )
+
+    def install(self, spec, prefix):
+        mkdir(prefix.bin)
+        install("build/release/duckdb", prefix.bin)
```
Kind of sad that the `Makefile` doesn't have an install target.

teaguesterling replied:
This may become increasingly sad as the build process seems to be expanding to create more artifacts than just the binary. Might be worth addressing with them upstream.

alalazo commented:
@glentner Do you have any comment on this PR?

alalazo commented on var/spack/repos/builtin/packages/duckdb/package.py:
```diff
@@ -16,8 +16,106 @@ class Duckdb(CMakePackage):
     version("0.7.1", sha256="67f840f861e5ffbe137d65a8543642d016f900b89dd035492d562ad11acf0e1e")
+
+    depends_on("python@3:")
+    depends_on("cmake")
+    depends_on("gmake")
+    depends_on("ninja", when="+ninjabuild")
```
```suggestion
    depends_on("cmake", type="build")
    depends_on("gmake", type="build")
    depends_on("ninja", when="+ninjabuild", type="build")
```

teaguesterling committed:
Update var/spack/repos/builtin/packages/duckdb/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

teaguesterling committed:
Update var/spack/repos/builtin/packages/duckdb/package.py

Thank you for the better idiomatic suggestion. I'm still becoming familiar with Spack packaging.

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

teaguesterling committed:
Update var/spack/repos/builtin/packages/duckdb/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

teaguesterling committed:
Including suggestions and comments from review.

teaguesterling committed:
Fixing setup_build_environment handling

teaguesterling committed:
Merge pull request #2 from teaguesterling/duckdb-review-changes

Including suggestions and comments from review.

teaguesterling committed:
Fixing black formatting for deprecations

teaguesterling committed:
Merge branch 'duckdb-review-changes' into packages/duckdb-with-version

teaguesterling committed:
Formatting...

teaguesterling committed:
Trailing whitespace

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

glentner commented:
I'm probably not the best person to review the Spack-specific build configuration. I will say this PR combines a lot of unrelated things, it might have been better to keep them separate. But that's not too important I don't think. I'm trying the latest commit and the build is failing for me (only default variants).

teaguesterling commented:
> I'm probably not the best person to review the Spack-specific build configuration. I will say this PR combines a lot of unrelated things, it might have been better to keep them separate. But that's not too important I don't think. I'm trying the latest commit and the build is failing for me (only default variants).

Do you have any information about the failure to build? I was able to test it today on 3 different systems successfully: 2 linux and 1 OSX and would like to chase down the error.

We could simplify this if it's causing stability issues (say, start with just the `patch` section to fix the version injection). For my purposes getting the in-tree extensions and out-of-tree extensions working was the original goal of these changes. The Makefile/CMake stuff was all about trying to follow the upstream build guidance. It seems that ended up getting a good deal more involved.

teaguesterling committed:
Merge branch 'develop' into packages/duckdb-with-version

EbiArnie commented:
Still builds and works for me with default variants. Built on Ubuntu Linux 22.04, gcc 11 from yesterday's develop@0543710.

alalazo commented:
Giving @glentner some time to post a reproducer.

alalazo commented:
@glentner I'm going ahead and merge this. Feel free to @teaguesterling and me as soon as you post an issu with a reproducer.
